### DISCUSSION TITLE
integration within Homey

### DISCUSSION THREAD

**Discussion by Aybert59** [mrbungle64/ecovacs-deebot.js#126] [Show and tell]
Dear sir

just to let you know, that I started to integrate your library within the Homey ecosystem

https://homey.app/fr-fr/app/com.messiant.ecovacs/Ecovacs/
https://github.com/Aybert59/com.messiant.ecovacs

**Comment by mrbungle64** [comment]
Hi @Aybert59 

that's great. Thanks for the information  👍🏻
Do not hesitate to contact me if you have any questions.

**Reply by Aybert59** [reply]
Hi I notice that version 0.7.2 gained 10 MB (without optional canvas).Any mean to remove additional optional packages ?
thank you

**Reply by mrbungle64** [reply]
I'm using axios as http client since version 0.7.2
Why are 10 MB too much?

**Reply by mrbungle64** [reply]
Canvas is the only optional dependency 😉 

```json
"dependencies": {
    "axios": "~0.26.0",
    "ltx": "~3.0.0",
    "mqtt": "~4.3.6",
    "simple-xmpp": "~2.3.1",
    "@xmldom/xmldom": "~0.8.1",
    "lzma": "~2.3.2",
    "uniqid": "~5.4.0"
  },
  "optionalDependencies": {
    "canvas": "~2.9.0"
  },
```

**Reply by Aybert59** [reply]
Thank you for the quick answer
Actually Homey recommends an app to be under 30MB size. With 0.7.1 I was at 15MB, now I'm at 25MB. 
Not yet critical anyway. Thank you

**Reply by mrbungle64** [reply]
* "axios" is used as http client
* "ltx" is used as XML builder
* "mqtt" is a client library for the MQTT protocol (necessary for most models)
* "simple-xmpp" is a client library for the XMPP protocol (necessary for legacy models)
* "xmldom" is used as XML parser
* "lzma" is used for converting compressed data
* "uniqid" is used for creating a unique id for the device

**Reply by Aybert59** [reply]
oh so sorry apparently some kind of misconfiguration happened where both versions were taken into account. Some cleaning and refresh for start, back to about 15MB

**Comment by mrbungle64** [comment]
@Aybert59 @cactusss

Why are you using this while loop instead of `setInterval()`?

```js
    const delay = ms => new Promise(resolve => setTimeout(resolve, ms));

    while (true) {
        this.log('syncing...');
        await this.syncStatus(this.vacbot)
              .catch(error => {
				        this.log('Sync error :', error);
			        });   
        await delay(SYNC_INTERVAL);
    }
```

@cactusss, you maybe already changed it. But the latest code you sent me used that while loop too.

Btw.: You don't have to poll for `GetBatteryState` (only once for initialization)

**Reply by Aybert59** [reply]
Hello
the initial purpose of that loop is to sync some status indicators regularly (like battery state). I understand you are rewriting this by managing the events which is more appropriate, as the robot device is verbose enough.
this is a template I use in other applications, for dumb devices, where there's no other way to update a status than polling an API regularly.
then the way I do it was grabbed from the homey developer forum, there was discussions on how to poll a device the more efficient and robust way. I tried setInterval but it was not more robust, with some experience (in an other app, 'syncStatus' triggers an http request, which is not handled properly when there is not response, that loop is just crashing).

**Reply by mrbungle64** [reply]
Hi @Aybert59 

I think the disadvantage of a while loop is that you have to implement a termination condition manually.
If you use `setInterval()` you can use `clearInterval()` to terminate that interval.

```js
this.getStatesInterval = setInterval(() => {
      this.vacbotGetStatesInterval();
}, this.pollingInterval);
```

```js
clearInterval(this.getStatesInterval);
```

**Comment by ghost** [comment]
Yes, I changed it already.

I think the reason @Aybert59 did it this way because in a setInterval the command `await this.syncStatus(this.vacbot)` gives an error.
I solved it by adding a `app = this` in the onInit() part, and then do a `await app.syncStatus(app.vacbot)`. I know this is a completly wrong way (and I don't use it anymore) but I also wouldn't know how to fix is. Any hints, as we (or at least I) are no guru's at all 😊
